### PR TITLE
Export Application view log

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Adresses
https://issues.redhat.com/browse/ODC-6401

# Screenshot
Button-jump-fix
button-jump-fix

Tooltip and log
export-view-log-btn

# Tests
Will add soon

# Browser conformance
Chrome

### CODE REVIEW

File: frontend/packages/topology/src/components/export-app/ViewLogButton.tsx
Comment (debsmita1) [reply frontend/packages/topology/src/components/export-app/ViewLogButton.tsx:60]: and shouldn't this be a link instead of a button ? cc @jerolimov

**Comment by debsmita1** [comment]
Please add some tests

File: frontend/packages/topology/src/components/export-app/ViewLogButton.tsx
Comment (jerolimov) [reply frontend/packages/topology/src/components/export-app/ViewLogButton.tsx:60]: Yep, it would be great to allow the user to ctrl/cmd + click / right-click open in a new tab so that it is not required to leave the topology.

File: frontend/packages/topology/src/components/export-app/ViewLogButton.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/ViewLogButton.tsx:50]:
```suggestion
  const viewLog = (event: React.MouseEvent<HTMLElement>) => {
    // Do nothing in this browser tab if the user pressed ctrl+click on the button
    if (isModifiedEvent(event)) return;
    history.push(path);
    onViewLog?.();
  };
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
I would recommend merging `ExportViewLogButton` and `ViewLogButton` as they are really tightly coupled.

Added also some other comments about using useMemo instead of settings a new state. Please take a look and let me know if you have another opinion or any other questions.

File: frontend/packages/topology/src/components/export-app/ExportApplication.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportApplication.tsx]: added a TIMEOUT const. Will raise a bug to use these notation in other places for TOASTs.

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx]: Using memoized data instead of effects

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx]: merged the components together

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx]: updated to fetch using jobname

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:24]: Please check also that the button shows "View logs" and calls the onClick method when clicked.

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:91]:
```suggestion
  return path ? linkedButton : disabledButton;
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Added another small change request to remove `isFlexItem` prop and replace it by just wrap the button. We discussed `isCreating` issue already on Slack and I added another hint for the button tests.

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:12]: There is an issue with the approach mentioned. I started off with that. But the problem arises when the job is being created or not ready and the `ExportViewLogButton` renders `null` There is a mismatch of gap between the buttons in the modal.

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: Updated tests

File: frontend/packages/topology/src/components/export-app/ExportApplication.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportApplication.tsx]: I have simplified it to be the same text `Export Application` as getting some of the data requires a lot of change in the utils.

File: frontend/packages/topology/src/components/export-app/ExportApplicationModal.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportApplicationModal.tsx]: removed FlexItem

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:17]: IMHO this type "extraction" makes the complete test less readable and is not necessary, you can just add namespace="test" to your react elements below.

I recommend to drop this:

```suggestion
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:21]: Drop this as well:

```suggestion
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:57]: And keep it simple and set the required props here:

```suggestion
        <ExportViewLogButton namespace="namespace" />
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:73]: Updating global is also more vulnerable to changes in other tests. Also if this is not an issue because you reset this here in `beforeEach`, but when it's easy to avoid this we should use simple local variables and props.

```suggestion
    const viewLogCallback = jest.fn();
    // mount required for using provider
    const logButtonWrapper = mount(
      <Provider store={store}>
        <ExportViewLogButton namespace="test" onViewLog={viewLogCallback} />
      </Provider>,
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:84]:
```suggestion
        <ExportViewLogButton namespace="test" />
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:121]:
```suggestion
        <ExportViewLogButton namespace="test" />
```

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (jerolimov) [inline frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx:154]:
```suggestion
        <ExportViewLogButton namespace="test" />
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
In general I'm fine with this solution. I just added two more comments:

1. The button "jumps" between disabled and enabled state.
2. The tests could be simplified a little bit

You also need to get feedback from UX.

/cc @beaumorley @serenamarie125

Beau, Serena, are you fine with this new button? :)

File: frontend/packages/topology/src/components/export-app/ExportApplication.scss
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportApplication.scss]: removed it

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: updated

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: updated

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: updated

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: updated

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: updated

File: frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/__tests__/ExportViewLogButton.spec.tsx]: removed

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:77]: fixed this 
button-jump-fix

**Comment by beaumorley** [comment]
@abhinandan13jan Looks close. However, were you able to remove the extra space around the "View logs" link (mentioned above by Christoph.  Maybe that is fixed when the jumping is fixed? It is hard to see so I cannot tell. Also where you able to fix the jumping on both the toast and the modal? I see it is fixed in the gif for the modal but not the toast. Thank you.

**Comment by abhinandan13jan** [comment]
@beaumorley @jerolimov 
Please ignore the initial screenshot those are to just portray the view log behaviour. Current button behaves as follows
button-jump-fix

**Comment by christianvogt** [comment]
I won't hold up the PR for this as UX is aware of the current design. Just sharing my thought.

Patternfly provides variations for their alerts. Alert actions typically go in the action toolbar of the alert which is below the content. There is a variation with a link in the content area but there is no guidance from PF as to when to use one vs the other. I feel like the action toolbar is the most appropriate as the action is not part of the flow in messaging (ie. it's not integrated into the sentence) 

image

**Comment by jerolimov** [comment]
Propagating Docs & PX ack labels from Epic [ODC-6296](https://issues.redhat.com/browse/ODC-6296)

/label docs-approved
/label px-approved

**Review by jerolimov — APPROVED** [review]
Tests :tada: 

No custom css anymore :+1: 

I agree with Christian and I'm also not a UX fan of the link button between other buttons, but I think we can merge this and reorder/realign buttons if needed.

Tested it locally, works fine. I will test it later also on a cluster bot instance and I will add the qe-approved label after that as well.

But can you please remove the now unused prop `additionalClassName` ?

/approve

**Comment by abhinandan13jan** [comment]
@jerolimov removed the prop

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (christianvogt) [inline frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:37]: Be careful when using ternary operators. In this case `useK8sWatchResource` expects param of type `WatchK8sResource | null` but `podResource` doesn't match this type. Looks like typescript type inference doesn't properly acknowledge that `podResource` may be `false` due to `jobLoaded`.

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (christianvogt) [inline frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:8]:
```suggestion
import { isModifiedEvent } from '@console/shared';
```

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (christianvogt) [inline frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:47]: Using a memo typically isn't necessary for string computations as the value isn't heavy weight to compute.
Furthermore the `podData` object is a very broad object to have as a dependency and therefore will cause the path to be recomputed more often than needed since you really just need the name and namespace.

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (jerolimov) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx:37]: Yeah, explicit `job && jobLoaded ? { ... } : null` would make this more typesafe.

**Comment by jerolimov** [comment]
Tested it on a cluster. Had some issues with triggering the export twice, but we addressed this issue already on Slack.

/label qe-approved

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx]: removed Memo

File: frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx
Comment (abhinandan13jan) [reply frontend/packages/topology/src/components/export-app/ExportViewLogButton.tsx]: returning null incase object cannot be formed

**Pull request merged**
